src/vision_overhaul.cpp

```diff
@@ -0,0 +1,455 @@
+/* General design: All characters have a vision array that takes NONE, IS_NATURAL, IS_ARTIFICIAL.
+   Vision-using code calls functions from here to see what level of vision they have.
+   We also rip out the vision mods from the current penalty function and put them here for centralization. */
+
+#include <map>
+
+#include "structs.h"
+#include "comm.h"
+#include "constants.h"
+
+bool _vision_prereqs_are_valid(struct char_data *ch, int type, const char *function_name);
+bool _vision_bit_is_valid(int bit, const char *function_name);
+
+// Do we have a given type of vision at all?
+bool has_vision(struct char_data *ch, int type, bool staff_override) {
+  if (!_vision_prereqs_are_valid(ch, type, "has_vision"))
+    return FALSE;
+
+  return ch->points.vision[type].HasAnythingSetAtAll();
+}
+
+// Do we have a given type of vision in a natural form? (Used for vision penalties)
+bool has_natural_vision(struct char_data *ch, int type) {
+  if (!_vision_prereqs_are_valid(ch, type, "has_natural_vision"))
+    return FALSE;
+
+  // Spell vision specifically functions as if cybernetic, so it doesn't qualify as natural.
+  return ch->points.vision[type].AreAnySet(VISION_BIT_IS_NATURAL, VISION_BIT_IS_ADEPT_POWER, ENDBIT);
+}
+
+// Do we have a given type of vision in a form we can cast with?
+bool has_castable_vision(struct char_data *ch, int type) {
+  if (!_vision_prereqs_are_valid(ch, type, "has_castable_vision"))
+    return FALSE;
+
+#ifndef DIES_IRAE
+  // Short circuit: If we're not using castable-vision restrictions, bail out.
+  return TRUE;
+#endif
+
+  // You can never cast through ultrasound (M&M p18).
+  if (type == VISION_ULTRASONIC)
+    return FALSE;
+
+  // You can always cast through natural vision.
+  if (has_natural_vision(ch, type))
+    return TRUE;
+
+  // Implanted cybernetics allow spellcasting per SR3 p181.
+  if (ch->points.vision[type].IsSet(VISION_BIT_FROM_IMPLANTS))
+    return TRUE;
+
+  // All other vision types block spellcasting. This specifically includes:
+  // - Electronic means of vision (SR3 p181)
+  // - Spell vision (same page)
+  // - Adept powers (adepts can't cast)
+  // - Equipment-based vision that you don't already have (LL/IR sensors are still electronic imaging)
+  return FALSE;
+}
+
+void remove_racial_vision_due_to_eye_replacement(struct char_data *ch) {
+  switch (GET_RACE(ch)) {
```

Comment:
This needs to handle the Cat's Eye bioware case, which grants natural vision even though it's not technically natural. It would be lost under the current code.